module/version_reconfigure_test.go

```diff
@@ -0,0 +1,235 @@
+			client.WithDialOptions(rpc.WithForceDirectGRPC()),
+			client.WithDisableSessions(), // TODO(PRODUCT-343): add session support to modules
+		)
+		dialCancel()
+		if !errors.Is(err, context.DeadlineExceeded) {
+			return rc, err
+		}
+		select {
+		case <-connectCtx.Done():
+			return nil, connectCtx.Err()
+		default:
+		}
+	}
+}
+
+// TODO(pre-merge) refactor this somewhat-shared function into a test_helpers esque file.
+func modifyCfg(t *testing.T, cfgIn string, logger golog.Logger) (string, string, error) {
+	p, err := goutils.TryReserveRandomPort()
+	if err != nil {
+		return "", "", err
+	}
+	port := strconv.Itoa(p)
+
+	cfg, err := config.Read(context.Background(), cfgIn, logger)
+	if err != nil {
+		return "", "", err
+	}
+	cfg.Network.BindAddress = "localhost:" + port
+	output, err := json.Marshal(cfg)
+	if err != nil {
+		return "", "", err
+	}
+	file, err := os.CreateTemp(t.TempDir(), "viam-multiversionmodule-config-*")
+	if err != nil {
+		return "", "", err
+	}
+	cfgFilename := file.Name()
+	_, err = file.Write(output)
+	if err != nil {
+		return "", "", err
+	}
+	return cfgFilename, port, file.Close()
+}
+
+func TestValidationFailureDuringReconfiguration(t *testing.T) {
+	logger, logs := golog.NewObservedTestLogger(t)
+
+	// Use a valid config
+	cfgFilename, port, err := modifyCfg(t, utils.ResolveFile("module/multiversionmodule/module.json"), logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	serverPath, err := testutils.BuildTempModule(t, "web/cmd/server/")
+	test.That(t, err, test.ShouldBeNil)
+
+	server := pexec.NewManagedProcess(pexec.ProcessConfig{
+		Name: serverPath,
+		Args: []string{"-config", cfgFilename},
+		CWD:  utils.ResolveFile("./"),
+		Log:  true,
+	}, logger)
+
+	err = server.Start(context.Background())
+	test.That(t, err, test.ShouldBeNil)
+	defer func() {
+		test.That(t, server.Stop(), test.ShouldBeNil)
+	}()
+
+	rc, err := connect(port, logger)
+	test.That(t, err, test.ShouldBeNil)
+	defer func() {
+		test.That(t, rc.Close(context.Background()), test.ShouldBeNil)
+	}()
+
+	// Assert that motors and base were added.
+	_, err = rc.ResourceByName(generic.Named("generic1"))
+	test.That(t, err, test.ShouldBeNil)
+
+	// Assert that there were no validation or component building errors
+	test.That(t, logs.FilterMessageSnippet(
+		"modular config validation error found in resource: generic1").Len(), test.ShouldEqual, 0)
+	test.That(t, logs.FilterMessageSnippet("error building component").Len(), test.ShouldEqual, 0)
+
+	// Read the config, swap to `run_version2.sh`, and overwrite the config, triggering a reconfigure where `generic1` will fail validation
+	cfg, err := config.Read(context.Background(), cfgFilename, logger)
+	test.That(t, err, test.ShouldBeNil)
+	cfg.Modules[0].ExePath = utils.ResolveFile("module/multiversionmodule/run_version2.sh")
+	newCfgBytes, err := json.Marshal(cfg)
+	test.That(t, err, test.ShouldBeNil)
+	os.WriteFile(cfgFilename, newCfgBytes, 0o644)
```

Comment:
Random thought: I'm curious if we've ever had some flakey test failures due to this overwriting a file in place not being atomic w.r.t to the module watching for a file change.

Reply:
That is a good point. Though, I dont think I've seen a test failure yet that could be attributed to that.